### PR TITLE
ci: pre-merge gate (size, secret-refs, major-bump test enforcement)

### PR DESCRIPTION
**Pull request by joshsmithxrm**
## Summary

Implements **PR-E** of the v1-launch retro (item #11): a pre-merge gate that enforces three rules on every PR before merge to `main`. Each rule retroactively catches a real failure that slipped through:

| Rule | Catches | Bypass marker |
|------|---------|---------------|
| 1 — PR size guardrail | PR #792 (131 files / 7.5K LoC merged unreviewed) | `[size-waived: <reason>]` |
| 2 — Workflow secret-ref drift | PR #797 / ppds-docs#15 (`AUDIT_REPO_TOKEN` referenced, didn't exist) | `[secret-ref-allow: <NAME>]` |
| 3 — Major-bump test enforcement | PR #806 (vite 5 -> 8, two majors, only `check-changes` ran) | none — fix the cause, don't waive |

All three rules are **ENFORCING from day one** — no soft-warn period (per retro recommendation). Bypass markers are case-sensitive in the PR title or body.

> **Do not auto-merge.** Orchestrator marks ready after CI.

## What's added

- `.github/workflows/pre-merge-gate.yml` — runs on `pull_request` (opened, synchronize, reopened, ready_for_review, edited). Each rule is its own job for individual diagnosability; an aggregate `Pre-Merge Gate` job is what branch protection should require.
- `scripts/ci/__init__.py`
- `scripts/ci/check_pr_size.py` — Rule 1 (>50 files OR >2000 LoC)
- `scripts/ci/check_workflow_secrets.py` — Rule 2 (regex-extracts `secrets.X` / `vars.X` from changed workflow files; compares against `gh secret list --json name` / `gh variable list --json name`)
- `scripts/ci/check_major_bump_tested.py` — Rule 3 (uses `classify_pr` from `scripts/dependabot/classify.py` (PR #814) for major detection; checks `gh pr checks --json name,state` for `test` job state)
- `tests/ci/__init__.py` and three `test_*.py` files (72 unit tests, all `gh` calls mocked)
- `docs/MERGE-POLICY.md` — new `## Pre-Merge Gate Rules` section

### Rule 1 — PR size (`scripts/ci/check_pr_size.py`)

Blocks merge when changed files > 50 OR additions+deletions > 2000.
Bypass: `[size-waived: <reason>]` in title or body. Empty / whitespace-only reasons rejected. Case-sensitive (deliberate copy-paste required).

### Rule 2 — Workflow secret-refs (`scripts/ci/check_workflow_secrets.py`)

For PRs touching `.github/workflows/*.{yml,yaml}`:
1. Regex-extracts `${{ secrets.X }}` and `${{ vars.X }}` (no YAML required since these are string substitutions).
2. Lists repo-level secret/variable **names** (not values) via `gh secret list --json name` and `gh variable list --json name` — `GITHUB_TOKEN` is sufficient, no elevated PAT.
3. Blocks if any reference is missing.

Built-in `GITHUB_TOKEN` is always considered present. Environment-scoped / org-level secrets are not enumerated by repo-level listing — use the bypass.

Bypass: `[secret-ref-allow: <NAME>]` in title or body, repeated per missing name.

### Rule 3 — Major-bump test enforcement (`scripts/ci/check_major_bump_tested.py`)

For dependabot PRs (label `dependencies` OR author `dependabot[bot]` / `app/dependabot`):
- Detects major bumps via `classify_pr` from `scripts/dependabot/classify.py` (PR #814).
- Requires the actual `test` job (not the path-filter `check-changes` skip-status) to have RUN AND PASSED.
- Blocks if `test` is `SKIPPED` / `FAILURE` / pending / absent.

**No bypass.** Fix by pushing an empty commit, re-running all jobs, or expanding path filters.

## Test plan

- [x] `python -m pytest tests/ci/ -v` — 72 passed in 0.07s (mocked, no network)
- [x] Rule 1 CLI smoke: under-limits returns 0; over-limits returns 1; with valid waiver returns 0
- [x] Workflow YAML parses (`yaml.safe_load`)
- [x] All three Python scripts compile (`python -m py_compile`)
- [ ] First real PR after merge will exercise the gate end-to-end against repo's actual `gh secret list`
- [ ] Optional: branch protection updated to require `Pre-Merge Gate` aggregate check (orchestrator decides when)

## Notes / tradeoffs

- **Bypass marker location:** title OR body, case-sensitive, regex-anchored — chose case-sensitivity to prevent accidental triggers from prose (`"Size waived this concern in slack..."`). Documented in `docs/MERGE-POLICY.md`.
- **Secret listing without elevated permissions:** `gh secret list --json name` and `gh variable list --json name` only return names, not values, and work with the default `GITHUB_TOKEN`. No PAT or org-level token required. Limitation: env-scoped and org-level secrets are not enumerated — the `[secret-ref-allow: <NAME>]` bypass exists for those legitimate cases.
- **Rule 3 dependency:** depends on `scripts/dependabot/classify.py` from PR #814 (now in main). The script imports `classify` directly (no copy of the regex), so any future changes to `_BUMP_TITLE_RE` / `classify_update_type` automatically apply.
- **Rule 3 grouped bumps:** `classify_pr` returns Group B / `update_type=unknown` for grouped bumps, so they don't fire Rule 3. The `/dependabot-triage` skill handles grouped-bump escalation.
- **Why a separate aggregate job:** branch protection requires named checks. Three rule jobs + one aggregate gives diagnosability without forcing branch protection to enumerate all three.

Refs: v1-launch retro item #11.

Co-Authored-By: Claude Opus 4.7 (1M context) <noreply@anthropic.com>

### CODE REVIEW

File: scripts/ci/check_workflow_secrets.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+        "body": data.get("body", "") or "",
+        "workflow_files": workflow_files,
+    }
+
+
+def fetch_repo_secrets() -> set[str]:
+    """List repo-level secret names via `gh secret list`."""
+    raw = _run_gh(["secret", "list", "--json", "name"])
+    data = json.loads(raw) if raw.strip() else []
+    return {item["name"] for item in data if "name" in item}
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply scripts/ci/check_workflow_secrets.py]: Fixed in 4cfa9d6a695489ca394f63c5c971c4d0e2140b30. `fetch_repo_secrets` now uppercases every name and uses `(item.get("name") or "").upper()` to null-coalesce. Combined with uppercasing extracted refs (HIGH 3106710728), the inventory comparison is now genuinely case-insensitive — `secrets.PPDS_TOKEN` matches a `gh secret list` entry of `ppds_token`. New test `TestInventoryFetchers::test_fetch_repo_secrets_uppercases_names` and `test_fetch_repo_secrets_handles_null_name_field` cover both behaviors.

File: scripts/ci/check_workflow_secrets.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+
+
+def fetch_repo_variables() -> set[str]:
+    """List repo-level variable names via `gh variable list`."""
+    raw = _run_gh(["variable", "list", "--json", "name"])
+    data = json.loads(raw) if raw.strip() else []
+    return {item["name"] for item in data if "name" in item}
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply scripts/ci/check_workflow_secrets.py]: Fixed in 4cfa9d6a695489ca394f63c5c971c4d0e2140b30. Same fix applied to `fetch_repo_variables` for symmetry: `(item.get("name") or "").upper()`. Tested by `TestInventoryFetchers::test_fetch_repo_variables_uppercases_names` and the case-insensitive var integration test `test_var_ref_case_insensitive_match` (workflow refs `vars.MyVar`, inventory has `MYVAR` — must pass).

File: scripts/ci/check_workflow_secrets.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+
+def extract_refs(workflow_text: str) -> tuple[set[str], set[str]]:
+    """Extract (secret_names, var_names) referenced in a workflow file.
+
+    Pure regex — ``yaml`` parsing isn't required because the references are
+    string-substituted before YAML semantics; regex catches them in any
+    YAML position (key, value, multi-line scalar).
+    """
+    secrets = set(SECRET_REF_RE.findall(workflow_text))
+    variables = set(VAR_REF_RE.findall(workflow_text))
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply scripts/ci/check_workflow_secrets.py]: Fixed in 4cfa9d6a695489ca394f63c5c971c4d0e2140b30. `extract_refs` now does `{s.upper() for s in SECRET_REF_RE.findall(...)}` and the same for vars. New test `TestExtractRefs::test_secret_refs_normalized_to_uppercase` confirms `secrets.lowercase_secret` extracts as `LOWERCASE_SECRET` and `vars.MixedCase_Var` as `MIXEDCASE_VAR`. Combined with the inventory fix, this closes the case-sensitivity false-positive that was failing the gate on itself.

File: scripts/ci/check_workflow_secrets.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+    return secrets, variables
+
+
+def find_allow_markers(title: str, body: str) -> set[str]:
+    """Return the set of names allowed via ``[secret-ref-allow: NAME]`` markers."""
+    allowed: set[str] = set()
+    for blob in (title or "", body or ""):
+        for m in ALLOW_MARKER_RE.finditer(blob):
+            allowed.add(m.group(1))
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply scripts/ci/check_workflow_secrets.py]: Fixed in 4cfa9d6a695489ca394f63c5c971c4d0e2140b30. `find_allow_markers` now does `allowed.add(m.group(1).upper())` so the bypass marker is matched case-insensitively against the (now also uppercase) referenced names. New test `TestCheckWorkflowSecrets::test_allow_marker_case_insensitive_match` proves a workflow with `secrets.MyLegitSecret` is bypassed by `[secret-ref-allow: mylegitsecret]`. Existing uppercase marker tests (`test_marker_in_title`, `test_marker_in_body`, etc.) still pass.

File: scripts/ci/check_major_bump_tested.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+def fetch_pr_checks(pr_number: int) -> list[dict]:
+    """Return the list of checks for the PR (gh pr checks --json)."""
+    raw = _run_gh(["pr", "checks", str(pr_number), "--json", "name,state"])
+    return json.loads(raw) if raw.strip() else []
+
+
+def is_dependabot_pr(pr: dict) -> bool:
+    """True if the PR is dependabot-originated."""
+    labels = {lbl.get("name", "").lower() for lbl in pr.get("labels", [])}
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply scripts/ci/check_major_bump_tested.py]: Fixed in 4cfa9d6a695489ca394f63c5c971c4d0e2140b30. `is_dependabot_pr` now uses `(lbl.get("name") or "").lower()` and iterates over `(pr.get("labels") or [])` so a null label name or null labels list won't TypeError. New tests `TestIsDependabotPr::test_null_label_name_treated_as_empty` and `test_null_labels_list_treated_as_empty` cover both shapes.

File: scripts/ci/check_major_bump_tested.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+    """Return (passed, message). Passes iff a required test job both ran and succeeded.
+
+    States we honor (per gh pr checks --json):
+      - SUCCESS / pass  -> ran and passed
+      - SKIPPED         -> did NOT run, fail the rule
+      - PENDING / IN_PROGRESS / QUEUED -> still running, fail (PR isn't ready)
+      - FAILURE / CANCELLED / TIMED_OUT / ACTION_REQUIRED / NEUTRAL -> fail
+      - missing entirely -> fail
+    """
+    by_name = {c.get("name", ""): (c.get("state", "") or "").upper() for c in checks}
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply scripts/ci/check_major_bump_tested.py]: Fixed in 4cfa9d6a695489ca394f63c5c971c4d0e2140b30. `check_test_job_ran` now does `{(c.get("name") or ""): (c.get("state") or "").upper() for c in checks}` so a null name or null state from `gh pr checks` is null-coalesced. New tests `TestCheckTestJobRan::test_null_check_name_does_not_crash` (still finds the real `test` job) and `test_null_check_state_treated_as_pending` cover both.

File: scripts/ci/check_pr_size.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+def check_pr_size(pr: dict) -> tuple[bool, str]:
+    """Apply the size rule. Returns (passed, message).
+
+    `pr` must have keys: additions, deletions, changedFiles, title, body.
+    """
+    try:
+        additions = int(pr.get("additions", 0))
+        deletions = int(pr.get("deletions", 0))
+        files = int(pr.get("changedFiles", 0))
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply scripts/ci/check_pr_size.py]: Fixed in 4cfa9d6a695489ca394f63c5c971c4d0e2140b30. All three size fields now use `int(pr.get("key") or 0)` so an explicit `null` from `gh pr view` is treated as 0 instead of TypeError. New tests `TestMalformedInput::test_null_additions_treated_as_zero`, `test_null_deletions_treated_as_zero`, `test_null_changed_files_treated_as_zero`, and `test_all_null_size_fields_treated_as_zero` cover the matrix. The existing `test_non_numeric_additions_fails` still passes since `"abc" or 0 == "abc"` and `int("abc")` still raises ValueError.

File: scripts/ci/check_workflow_secrets.py [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+import re
+import subprocess
+import sys
+from typing import Iterable, Optional
+
+# Match ``${{ secrets.NAME }}`` and ``${{ vars.NAME }}`` allowing arbitrary
+# whitespace inside the ``${{ }}`` and around the dot. Names are
+# letters/digits/underscores per GitHub Actions spec.
+SECRET_REF_RE = re.compile(r"\$\{\{\s*secrets\.([A-Za-z_][A-Za-z0-9_]*)\s*\}\}")
+VAR_REF_RE = re.compile(r"\$\{\{\s*vars\.([A-Za-z_][A-Za-z0-9_]*)\s*\}\}")
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply scripts/ci/check_workflow_secrets.py]: Fixed in 4cfa9d6a695489ca394f63c5c971c4d0e2140b30. Both regexes now permit whitespace around the dot per the docstring contract: `SECRET_REF_RE = re.compile(r"\$\{\{\s*secrets\s*\.\s*([A-Za-z_][A-Za-z0-9_]*)\s*\}\}")` and the same shape for vars. New test `TestExtractRefs::test_secret_ref_with_whitespace_around_dot` verifies `${{ secrets . FOO }}` and `${{ vars  .  BAR }}` extract correctly. Existing `test_handles_arbitrary_whitespace_in_curly` still passes (no change to outer-curly whitespace behavior).

File: scripts/ci/check_workflow_secrets.py [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+
+
+def fetch_pr_metadata(pr_number: int) -> dict:
+    """Return PR title, body, and changed workflow files."""
+    raw = _run_gh([
+        "pr", "view", str(pr_number),
+        "--json", "title,body,files",
+    ])
+    data = json.loads(raw)
+    files = [f.get("path", "") for f in data.get("files", [])]
```
[the remark this answers is not in the document]
Comment (joshsmithxrm) [reply scripts/ci/check_workflow_secrets.py]: Fixed in 4cfa9d6a695489ca394f63c5c971c4d0e2140b30. `fetch_pr_metadata` now does `[f.get("path") or "" for f in (data.get("files") or [])]` so a null `files` array or null `path` field from `gh pr view` won't TypeError. New tests `TestInventoryFetchers::test_fetch_pr_metadata_handles_null_files_list` and `test_fetch_pr_metadata_handles_null_path_in_file` cover both edge cases.
